adsmp/app.py

```diff
@@ -160,3 +217,59 @@ def get_msg_status(self, msg):
                 return 'active'
         else:
             return 'unknown'
+        
+    
+
+    def reindex(self, solr_docs, batch_insert, batch_update, solr_urls):
+        """Sends documents to solr and to Metrics DB.
+        
+        :param: solr_docs - list of json objects (solr documents)
+        :param: batch_insert - list of json objects to insert into the metrics
+        :param: batch_update - list of json objects to update in metrics db
+        :param: solr_urls - list of strings, solr servers.
+        """
+        
+        out = solr_updater.update_solr(solr_docs, solr_urls, ignore_errors=True)
+        errs = [x for x in out if x != 200]
+        
+        if len(errs) == 0:
+            self._mark_processed(solr_docs)
+            self._insert_into_metrics_db(batch_insert, batch_update)
+        else:
+            # recover from erros by inserting docs one by one
+            for doc in solr_docs:
+                try:
+                    solr_updater.update_solr([doc], solr_urls, ignore_errors=False)
+                    self.update_processed_timestamp(doc['bibcode'])
+                except:
+                    failed_bibcode = doc['bibcode']
+                    self.logger.error('Failed posting data to %s\noffending payload: %s', solr_urls, doc)
+                    
+                    # when solr_urls > 1, some of the servers may have successfully indexed
+                    # but here we are refusing to pass data to metrics db; this seems the 
+                    # right choice because there is only one metrics db (but if we had many,
+                    # then we could differentiate) 
+                    
+                    batch_insert = filter(lambda x: x['bibcode'] != failed_bibcode, batch_insert)
+                    batch_update = filter(lambda x: x['bibcode'] != failed_bibcode, batch_update)
+                    
+            self._insert_into_metrics_db(batch_insert, batch_update)
+
+    
+    def _mark_processed(self, solr_docs):
+        bibcodes = [x['bibcode'] for x in solr_docs]
+        now = adsputils.get_date()
+        with self.session_scope() as session:
+            session.query(Records).filter(Records.bibcode == bibcodes).update({'processed': now})
+            session.commit()
+
+
+    def _insert_into_metrics_db(self, batch_insert, batch_update):
+        """Inserts data into the metrics DB."""
+        if not self._metrics_table:
+            raise Exception('You cant do this! Missing METRICS_SQLALACHEMY_URL?')
```

Comment:
if converted to log, then the pipeline would be running - but in unhealthy state - better to die fast (the same as above)